core/drivers/crypto/se050/core/ecc.c

```diff
@@ -0,0 +1,627 @@
+		free(key_bin.d);
+		free(key_bin.x);
+		return ret;
+	}
+
+	key_bin.curve = curve_tee2se050(key->curve);
+	st = se050_key_store_set_ecc_key_bin(se050_kstore, k_object, &key_bin,
+					     NULL);
+	free(key_bin.d);
+	free(key_bin.x);
+	free(key_bin.y);
+	if (st != kStatus_SSS_Success)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result shared_secret(struct ecc_keypair *private_key,
+				struct ecc_public_key *public_key,
+				void *secret, unsigned long *secret_len)
+{
+	struct se050_ecc_keypub key = { };
+	sss_status_t st = kStatus_SSS_Fail;
+	TEE_Result ret = TEE_SUCCESS;
+	size_t key_bits = 0;
+	size_t key_bytes = 0;
+	size_t x1_len = 0;
+	size_t y1_len = 0;
+	size_t x2_len = 0;
+	uint32_t kid = 0;
+
+	if (private_key->curve != public_key->curve)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	kid = se050_ecc_keypair_from_nvm(private_key);
+	if (!kid) {
+		EMSG("private key must be stored in SE050 flash");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	/* validate input parameters */
+	x1_len = crypto_bignum_num_bytes(private_key->x);
+	y1_len = crypto_bignum_num_bytes(private_key->y);
+	x2_len = crypto_bignum_num_bytes(public_key->x);
+
+	ret = ecc_get_key_size(public_key->curve, 0, &key_bytes, &key_bits);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	if (x1_len != y1_len || x1_len != key_bytes || x1_len != x2_len)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* prepare the public key (must be in raw format) */
+	ret = set_binary_data(public_key->x, x1_len, &key.x, &key.x_len);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	ret = set_binary_data(public_key->y, y1_len, &key.y, &key.y_len);
+	if (ret != TEE_SUCCESS) {
+		free(key.x);
+		return ret;
+	}
+
+	st = se050_ecc_gen_shared_secret(&se050_session->s_ctx, kid, &key,
+					 secret, secret_len);
+	free(key.x);
+	free(key.y);
+	if (st != kStatus_SSS_Success)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sign(uint32_t algo, struct ecc_keypair *key,
+		       const uint8_t *msg, size_t msg_len,
+		       uint8_t *sig, size_t *sig_len)
+{
+	sss_status_t st = kStatus_SSS_Fail;
+	sss_se05x_asymmetric_t ctx = { };
+	sss_se05x_object_t kobject = { };
+	TEE_Result res = TEE_SUCCESS;
+	size_t key_bytes = 0;
+	size_t key_bits = 0;
+
+	res = ecc_get_key_size(key->curve, algo, &key_bytes, &key_bits);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	/* se050 exports DER format */
+	if (*sig_len < (2 * key_bytes + DER_SIGNATURE_SZ)) {
+		*sig_len = 2 * key_bytes + DER_SIGNATURE_SZ;
+		res = TEE_ERROR_SHORT_BUFFER;
+		goto exit;
+	}
+
+	res = ecc_get_msg_size(algo_tee2se050(algo), &msg_len);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = se050_inject_keypair(&kobject, key, key_bytes);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	st = sss_se05x_asymmetric_context_init(&ctx, se050_session, &kobject,
+					       algo_tee2se050(algo),
+					       kMode_SSS_Sign);
+	if (st != kStatus_SSS_Success) {
+		sss_se05x_key_store_erase_key(se050_kstore, &kobject);
```

Comment:
Above you're using the `exit:` label for cleanup, but not here. What's different?

Reply:
an error on my part. we have our own implementation of pkcs11 that is capable of importing pre-burnt secure objects (keys and certificates) from the SE050 to the database....and while preparing this patch I made a mistake. fixing it (thanks!)
https://docs.foundries.io/latest/reference-manual/security/secure-elements/secure-element.050.html#importing-secure-objects-to-pkcs-11-tokens